### PR TITLE
Improve error recovery in type argument list

### PR DESCRIPTION
**Pull request by YairHalberstadt**
This tries to prevent entire methods/properties/fields/classes being mangled by the parser when a '>' token is missing in certain locations. 

For example:

```csharp
public class A : List<int
{
    public void M()
    {
        M();
        C.M();
    }
}

public class B
{
    
    public List<int M()
    {
        var x = 42;
    }
    
    public List<int F = 42;
    
    public List<int P {get;set;}
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABAzAAmwCZ8BBfEfAGQEtYAeGgOxgFgAoAbw/18IOwAWfAFkAFAEoefbuz7zRkgNzSF+AMIA6cRJVy+AXw5H2HPIRIAhDrPmr+1OjEYtFU/b1tr8ANwgJ8AA98AF58QSI9eRM7DwdaBmYYfAAxUPDI+3tzBOck/AAFfE4AczgYJShypRMTDiA==

The relevant issue is #24642
A previous closed PR is at https://github.com/dotnet/roslyn/pull/32124

### CODE REVIEW

**Comment by YairHalberstadt** [comment]
@CyrusNajmabadi 
I haven't refactored all List parsing code yet, as I would like to get a bit more familiar with the codeBase first :-), so I have just implemented a solution for this one case for now.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -3275,14 +3275,15 @@ private enum PostSkipAction
             SeparatedSyntaxListBuilder<TNode> list,
             Func<LanguageParser, bool> isNotExpectedFunction,
             Func<LanguageParser, bool> abortFunction,
-            SyntaxKind expected)
+            SyntaxKind expected,
+            bool resetOnAbort = false)
```
[the remark this answers is not in the document]
Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Sure. I just assumed an optional parameter would be better than changing 15 locations in the commit.

[the remark this answers is not in the document]
Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: done

**Comment by RikkiGibson** [comment]
@YairHalberstadt do you mind merging the latest master into this so we can get up-to-date CI?

**Comment by YairHalberstadt** [comment]
It doesn't seem to be picking up the latest commit. Closing and reopening

**Comment by Rekkonnect** [comment]
Jumped into this PR after browsing issues related to poor error recovery, why was this PR closed? My personal experience with poor error recovery issues has been slowly growing into a fear of slowing down my experience in writing code, while being prepared for the massive snowstorm of red squiggles, despite being fluent in the language and knowing exactly what the compiler is crying for.

**Comment by CyrusNajmabadi** [comment]
@Rekkonnect we take prs that improve error recovery.

**Comment by Rekkonnect** [comment]
That's good to hear, but I still wonder why this PR was closed. It looked like it was solving the specific issue. Was it not meeting expectations, did it fail to cover a case that required rebuilding the fix, were the conflicts too many to resolve? I'd be interested to begin improving error recovery on specific areas, but avoiding redoing the work.

**Comment by CyrusNajmabadi** [comment]
>  I'd be interested to begin improving error recovery on specific areas, but avoiding redoing the work.

In general, list parsing recovery is non-trivial. You need to make sure that your new behavior doesn't only make some recovery cases better, it should also not make other ones worse.  And it needs to do this without exploding the parser in complexity there, and also not becoming a problem for future work for future language changes. 

The best thing to do is start with a ton of cases for that area, with anywhere from one to a few messed up tokens (both added and missing). Record what the current parser does with those. Then make the change and show how it's an improvement.

**Comment by Rekkonnect** [comment]
Will keep in mind, it seems like an interesting area to explore and improve

**Pull request closed without merging**